[Refactor] 예약 기능 리팩토링 - 결제 전 슬롯 임시 선점 및 만료 관리 기능 도입

dbghwns123 opened the pull request:
## 📌 관련 이슈
- close #261 

## 📝 변경 사항
### AS-IS
- 예약 검증 → 결제 → 슬롯 확보
- 결제 완료 후 예약 실패 가능 → 사용자 환불 필요

### TO-BE
- 예약 검증 시 슬롯 임시 선점 (hold = true, expiredAt = now + 5분)
- 결제 성공 후 슬롯 상태 확인 및 예약 확정 처리
- 만료된 슬롯은 스케줄러를 통해 hold = false, expiredAt = null로 초기화
- `findAndValidateHoldSlots()`에서 hold 상태 및 만료 여부 검증 강화

## 🔍 테스트
- [x] 테스트 코드 작성
- [x] API 테스트
- [x] 로컬 테스트

## 📸 스크린샷
- 결제 전 예약 가능 여부 확인 및 예약 슬롯 선점 (성공)
![image](https://github.com/user-attachments/assets/6842588a-bb39-4bfe-a026-1aac236e2f03)
(DB에 예약 슬롯 정상적으로 저장)
![image](https://github.com/user-attachments/assets/bd76676e-a7b8-4c95-a955-616c62fa0d69)

- 결제 전 예약 가능 여부 확인 및 예약 슬롯 선점 (실패) - 이미 예약한 객실
![image](https://github.com/user-attachments/assets/45e520e3-57e9-445b-a6fe-560bc18c0564)

- 결제 실패(존재하지 않는 결제 정보)
** 오류 메시지 로그: org.springframework.web.client.HttpClientErrorException$NotFound: 404 Not Found: "{"code":-1,"message":"존재하지 않는 결제정보입니다.","response":null}"

![image](https://github.com/user-attachments/assets/95310fcd-5102-496e-bd73-dfa9b91b59b7)


## ✅ 체크리스트
- [x] main/develop 브랜치가 아닌 feature 브랜치에서 작성했나요?
- [x] 코딩 컨벤션을 준수했나요?
- [x] 불필요한 코드는 제거했나요?
- [x] 주석은 충분히 작성했나요?
- [x] 테스트는 완료했나요?

## 🙋‍♂️ 리뷰어에게
이번 PR은 실사용 환경에서 발생 가능한 동시성 문제를 구조적으로 해결하기 위한 중요한 리팩토링입니다. 설계 구조, 예외 처리, 스케줄링 처리 등에서 개선사항이나 누락된 점이 있다면 피드백 부탁드립니다.

dbghwns123 committed:
feat: 예약 엔티티에 impUid 및 merchantUid 필드 추가

- 결제 연동을 위한 아임포트 결제번호(impUid)와 주문번호(merchantUid) 컬럼 추가
- 추후 결제 취소에도 쓰이기 때문에 추가

dbghwns123 committed:
feat: 예약 슬롯 동시성 제어를 위한 hold 및 expiredAt 필드 추가

- 결제 시도 중 슬롯 임시 점유(hold) 및 만료 시간(expiredAt) 관리 기능 구현을 위한 컬럼 추가
- 결제 진행 중 다른 사용자의 중복 예약 방지 목적

dbghwns123 committed:
feat: 다른 사용자가 결제중인(예약 슬롯의 hold = true) 객실 및 날짜인지 검증하는 checkRoomHoldStatus 메서드 추가

dbghwns123 committed:
feat: 예약 슬롯 임시 점유 확인을 위한 레포지토리 메서드 추가

- existsByRoomIdAndReservedDateBetweenAndHoldTrue 메서드 추가
- 특정 객실의 날짜 범위 내에서 임시 점유(hold)된 슬롯 존재 여부 확인 기능
- 결제 진행 중 중복 예약 방지 로직 구현을 위한 기반 마련

dbghwns123 committed:
feat: 결제 진행 중 임시 선점 상태 에러코드(ROOM_TEMPORARILY_HELD) 추가

dbghwns123 committed:
feat: 예약 슬롯 임시 선점(hold 및 expiredAt 설정) 기능 구현

- 결제 진행 전 기본적인 검증 후 예약 슬롯을 hold=true, expiredAt=now+5분으로 임시 점유하는 로직 추가
- 중복 예약 방지 및 동시성 제어 강화

dbghwns123 committed:
refactor: 메서드 명을 validateReservation에서 validateAndHoldSlots로 변경

- 메서드 역할을 명확히 드러내기 위해 이름 수정
- 예약 슬롯 검증 및 임시 선점(hold) 기능을 모두 포함함을 반영

dbghwns123 committed:
docs: 예약 슬롯 임시 선점 로직에 대한 주석 추가

dbghwns123 committed:
feat: hold 상태인 슬롯들을 조회하고 유효성 검증하는 findAndValidateHoldSlots 메서드 추가

- 결제 전 호출되는 validateAndHoldSlots 메서드의 내부 메서드인 holdReservationSlots 에서 생성된 선점된 예약 슬롯이 유효한지 검증하는 로직
- 각각의 예약 슬롯의 hold = false 또는 null 일 경우 예외 처리

dbghwns123 committed:
feat: 유효성 검증을 마친 예약 슬롯 예약 확정 처리하는 confirmReservationSlots 메서드 구현

- hold → reserved (IsReserved = true, Hold = false, ExpiredAt = null) 처리

dbghwns123 committed:
refactor: 사용하지 않는 예약 슬롯 예약 메서드(bookRoomForDates) 삭제

- 로직 변경에 따라 불필요해진 메서드 정리

dbghwns123 committed:
feat: 결제연동과 결제취소를 위해서 필요한 ImpUid, MerchantUid 필드가 Reservation 엔티티에 추가됨으로써 reservation 객체 생성 시 추가

dbghwns123 committed:
feat: 기존에는 주문번호를 랜덤으로 생성하여 반환했지만 결제 기능을 추가하면서 포트원에서 예약을 구분하는 고유 키로 사용하는 merchant_uid를 반환하도록 변경

dbghwns123 committed:
refactor: @Transactional(readOnly = true) → @Transactional로 변경

- 데이터 수정이 발생하는 로직에 맞게 트랜잭션 속성 수정

dbghwns123 committed:
fix: 만료된 hold 슬롯 유효성 검사 추가 및 예외 메시지 명확화

- 결제 시 슬롯 유효성 검사 시점에 hold=false, expiredAt=null 또는 expiredAt<now 인 경우 예외 처리
- ROOM_ALREADY_RESERVED → RESERVATION_SLOT_EXPIRED로 구분하여 에러 메시지 명확화

dbghwns123 committed:
feat: 예약 보류 시간 만료 에러코드(RESERVATION_SLOT_EXPIRED) 추가

dbghwns123 committed:
feat: 만료된 예약 hold 슬롯을 자동 초기화하는 스케줄러 추가

- hold=true && expiredAt < now 인 슬롯들을 주기적으로 초기화 (5분 간격)
- 해당 예약 슬롯들의 Hold = false, ExpiredAt = null 처리

dbghwns123 committed:
feat: 만료된 hold 상태의 예약 슬롯 조회 쿼리 추가

- ReservationSlotRepository에 findAllExpiredHoldSlots 메서드 및 JPQL 쿼리 구현
- expiredAt이 현재 시간보다 이전인 hold=true 슬롯 조회 기능

dbghwns123 committed:
test: 예약 슬롯 임시 선점(hold) 상태 예외 처리 테스트 추가

- checkRoomHoldStatus()에서 hold=true 슬롯 존재 시 ROOM_TEMPORARILY_HELD 예외 발생 검증

dbghwns123 committed:
test: hold가 false인 예약 슬롯 예외 처리 테스트 추가

- findAndValidateHoldSlots()에서 hold=false일 때 RESERVATION_SLOT_EXPIRED 예외 발생 검증

dbghwns123 committed:
test: expiredAt이 null인 예약 슬롯 예외 처리 테스트 추가

- findAndValidateHoldSlots()에서 expiredAt=null일 때 RESERVATION_SLOT_EXPIRED 예외 발생 검증

dbghwns123 committed:
test: expiredAt이 현재 시간보다 이전인 경우 예약 슬롯 예외 처리 테스트 추가

- findAndValidateHoldSlots()에서 expiredAt이 현재 시간보다 이전일 때 RESERVATION_SLOT_EXPIRED 예외 발생 검증

dbghwns123 committed:
test: 예약 등록 성공 테스트에서 불필요한 mock 제거 및 명시적 슬롯 검증 추가

- 기존 테스트 코드에서 실제 사용되지 않는 stubbing 제거 (UnnecessaryStubbingException 해결)
- 예약 슬롯(mock) 2건 생성 후 hold 상태와 expiredAt 설정 추가
- reservationSlotRepository.saveAll() 인자 검증을 List.of(slot1, slot2)로 명확히 수정
- ReservationResponse의 orderNumber에 대해 명시적 값 검증 추가
- 테스트 안정성을 높이기 위해 Reservation 객체에 impUid 및 merchantUid 설정 반영

기존 reservationService의 예약 흐름(findAndValidateHoldSlots → confirmReservationSlots)에 맞추어 테스트 데이터 구성 방식 개선

JJae-story commented on src/main/java/com/meongnyangerang/meongnyangerang/service/ReservationService.java:
```diff
@@ -168,44 +184,67 @@ private void checkRoomAvailability(Room room, LocalDate checkInDate, LocalDate c
+      reservationSlotRepository.saveAll(slots);
     } catch (OptimisticLockException e) {
       throw new MeongnyangerangException(ErrorCode.ROOM_ALREADY_RESERVED);
     }
```
reservationSlotRepository 에 saveAll 하는 걸 try ~ catch 로 감싸둔 이유는 동시성 문제로 예외 발생 시 예외를 던지려고 사용했던 거라서 제외하시면 될 것 같습니다.

JJae-story commented on src/main/java/com/meongnyangerang/meongnyangerang/service/ReservationService.java:
```diff
@@ -168,44 +184,67 @@ private void checkRoomAvailability(Room room, LocalDate checkInDate, LocalDate c
+      slot.setHold(true);
+      slot.setExpiredAt(LocalDateTime.now().plusMinutes(5));
+      reservationSlotRepository.save(slot);
+    }
```
현재 작성해주신 구조에서는 for문을 돌며 매번 `save()`를 호출하고 있어서, 아래와 같은 상황에서 `OptimisticLockingFailureException`이 발생할 수 있습니다:

- A 사용자가 1/1~1/3 사전 예약 API 호출 → 1/1, 1/2, 1/3 슬롯 순차 처리
- B 사용자가 1/2~1/4 사전 예약 API 호출 → 1/2, 1/3, 1/4 슬롯 순차 처리
- A와 B가 거의 동시에 1/2 슬롯을 생성 및 저장 시도 → 둘 다 `동일한 version` 상태에서 `save` 시도
- A가 먼저 저장에 성공하고 B가 실패하는 경우는 괜찮지만,
- 반대로 B가 먼저 저장하면 A는 이후 슬롯 저장 도중 예외가 발생하여 전체 요청이 실패할 수 있습니다.

이런 상황을 방지하기 위해, 이전 방식처럼 모든 슬롯을 처리한 후 `saveAll()`로 한 번에 저장하는 방식이 더 안정적일 것 같습니다.

JJae-story reviewed (commented):
수고하셨습니다!

dbghwns123 committed:
refactor: confirmReservationSlots 메서드에서 saveAll - try~catch문 삭제

- 이미 validateReservation() → createReservationAfterPayment() 로직 상에서 중복 예약/hold 여부를 체크했기 때문에 필요하지 않음

dbghwns123 committed:
refactor: 예약 슬롯 저장 방식 saveAll로 변경하여 낙관적 락 기반 동시성 제어 적용

- 기존에는 슬롯마다 개별적으로 save()를 호출했으나, 동시성 상황에서 OptimisticLockException이 발생할 수 있고, 일관성 있는 롤백 처리가 어려웠음.
- 이제 모든 슬롯을 수집한 후 saveAll()로 일괄 저장하도록 변경하여 트랜잭션 단위로 실패를 제어할 수 있도록 개선함.

Seokjin-K commented on src/main/java/com/meongnyangerang/meongnyangerang/service/ReservationService.java:
```diff
@@ -168,58 +184,87 @@ private void checkRoomAvailability(Room room, LocalDate checkInDate, LocalDate c
-      // 해당 날짜에 예약 슬롯을 조회하고 없으면 생성
       ReservationSlot slot = reservationSlotRepository
           .findByRoomIdAndReservedDate(room.getId(), date)
           .orElseGet(() -> new ReservationSlot(room, finalDate, false));
```
```.orElseGet(() -> new ReservationSlot(room, finalDate, false));``` 해당 부분에서 제 생각은:
만약 여러 사용자가 같은 날짜의 슬롯을 생성하고 DB에 저장하려고 한다면(해당 경우가 얼마나 발생할지는 모르지만) -> 가장 먼저 트랜잭션이 완료된 사용자의 데이터만 DB에 Insert -> 다른 사용자들은 ConstraintViolationException(이름이 맞는지 모르겠네요) 예외가 발생할 것 같은데 어떻게 생각하시나요?
만약 발생한다면 해당 예외에 대한 핸들링은 필요 없을까요?

**ConstraintViolationException 발생 이유** (이 또한 제 생각입니다)
현재 ReservationSlot은 room_id와 reserved_date에 Unique 유니크 제약 조건이 설정돼 있는 것으로 알고 있습니다.
때문에 여러 사용자가 같은 날짜의 슬롯을 생성하고 DB에 저장하려고 할 때 트랜잭션에서 ConstraintViolationException 예외를 발생시킬 것 같다는 생각이 듭니다.

호준님 생각은 어떠신가요?

dbghwns123 replied:
석진님 좋은 피드백 감사합니다! 말씀해주신 부분을 처음에는 인지하지 못했는데, 찾아보니 충분히 발생할 수 있는 시나리오라는 걸 알게 되었습니다.

먼저, 동시 다발적으로 동일한 예약 슬롯을 생성하려는 상황에서, 유니크 제약 조건에 의해 ConstraintViolationException 또는 DataIntegrityViolationException이 발생할 수 있다는 점을 새롭게 알게 되었습니다.
기존 코드에서는 OptimisticLockException만 처리하고 있었기 때문에, DataIntegrityViolationException예외도 추가하여 예외 처리에 반영했습니다.

여기서 ConstraintViolationException이 아닌 DataIntegrityViolationException만 catch한 이유는 다음과 같습니다:
DB에서 유니크 제약 조건 위반이 발생하면 JPA 내부에서는 ConstraintViolationException을 던지지만, Spring은 이를 내부에서 감싼 뒤 DataIntegrityViolationException으로 외부에 전달하기 때문에 실제 서비스 코드에서는 DataIntegrityViolationException만 catch하면 충분하다는 점을 확인했습니다.

또한, 이 예외 상황은 실제로는 **결제가 완료된 예약이 아닌, 다른 사용자가 해당 슬롯을 임시로 선점 중인 상태**이기 때문에, 사용자 입장에서는 "이미 예약된 객실입니다"보다는 "해당 기간은 다른 사용자가 결제 진행 중인 상태입니다. 잠시 후 다시 시도해 주세요."라는 메시지가 더 적절하다고 판단하여 에러코드도 ROOM_ALREADY_RESERVED → ROOM_TEMPORARILY_HELD로 변경하였습니다.

Seokjin-K replied:
**"DB에서 유니크 제약 조건 위반이 발생하면 JPA 내부에서는 ConstraintViolationException을 던지지만, Spring은 이를 내부에서 감싼 뒤 DataIntegrityViolationException으로 외부에 전달하기 때문에 실제 서비스 코드에서는 DataIntegrityViolationException만 catch하면 충분하다"**
해당 내용은 저도 처음 알게 되었네요. 감사합니다.

Seokjin-K reviewed (commented):
수고하셨습니다!

dbghwns123 committed:
fix: 동시성 이슈로 발생 가능한 중복 슬롯 생성 예외 처리 추가 (DataIntegrityViolationException 대응)

dbghwns123 committed:
fix: 예약 슬롯 충돌 시 ROOM_TEMPORARILY_HELD 에러코드로 변경

- 결제 전 단계의 예약 슬롯 충돌(OptimisticLock/DataIntegrity)을 ROOM_ALREADY_RESERVED → ROOM_TEMPORARILY_HELD로 변경
- "다른 사용자가 결제 진행 중"이라는 사용자 친화적 메시지로 더 명확한 안내 제공

Seokjin-K reviewed (approved):
수고하셨습니다!

dbghwns123 committed:
docs: 예약 사전 검증 API 주석을 상세히 수정

dbghwns123 committed:
docs: createReservationAfterPayment 메서드에 상세 주석 추가
